### PR TITLE
Add EngineeringTask

### LINKED ISSUE #16445: New EngineeringTask Tech Spec
# Adding an `EngineeringTask` to indicate that a Caseflow engineer is working on an appeal
**Drafter**: @yoomlam
**Discussion Meeting**:   July 14th, 2021 (during Caseflow Engineering Huddle)

## Context
Why are you creating this tech spec? 

The `AppealsWithNoTasksOrAllTasksOnHoldQuery` of the `StuckAppealsChecker` alerted on an appeal with no active task. Upon [investigation](https://dsva.slack.com/archives/CN3FQR4A1/p1625058496039500?thread_ts=1625049333.039300&cid=CN3FQR4A1), it turns out this is expected behavior for unrecognized appellants until a further workflow is implemented in Caseflow.  If an `EngineeringTask` was active on this appeal, there would not be an alert for this appeal.

Also, a new `EngineeringTask` would indicate to Caseflow users that engineers are working on the appeal, which would be informative and prevent false alarms for the Board that may see such appeals as being "stuck" in Tableau reports or similar. Also for reporting purposes, the duration that the `EngineeringTask` took can be deducted from the total time spent by the Caseflow user.

What information helps readers understand the rest of this tech spec? 

* https://github.com/department-of-veterans-affairs/caseflow/wiki/Tasks
* For reference, example task trees can be found from https://github.com/department-of-veterans-affairs/appeals-team/blob/master/Project%20Folders/Tasks/tasktrees/descr/tasks-overview.md#hearing-phase

A child `EngineeringTask` would block the parent task. In the case of the `StuckAppealsChecker` alert, a `BvaDispatchTask` would be on hold with an active `EngineeringTask` child, which blocks any progress on the parent `BvaDispatchTask` until the `EngineeringTask` is closed. Alternatively, an active `EngineeringTask` child could be created under the `RootTask` -- this would not block any specific user-worked task but would at least indicate that this appeal is being worked and is not stuck.

## Overview

There is benefit to adding a new `EngineeringTask`. It would keep Caseflow users aware of engineering work (including long-term Bat Team work) on an appeal, reduce false positives when checking for stuck appeals, and enable more accurate time-keeping when reporting on Caseflow users' time spent on an appeal.

Task `instructions` can be populated to further explain why an appeal has an `EngineeringTask`.

### Requirements and/or Acceptance Criteria

* indicate to Caseflow users that engineers are working on the appeal
* prevent false positives when checking for stuck appeals
* the `EngineeringTask` is assigned to a specific engineer when possible, otherwise it is assigned to the Caseflow user `css_id: "CSFLOW"`
* to indicate that we are not actively working on the appeal, the `EngineeringTask` can be put on hold, probably via a [`TimedHoldTask`](https://github.com/department-of-veterans-affairs/caseflow/wiki/Timed-Tasks#timedholdtask)

## Open Questions

* Do we want to show the `EngineeringTask` on the Case Details page? In the case timeline?
 

## Research Notes

# Tech Spec Process

 - [x] Tech spec drafter: Schedule time to discuss the tech spec with scrum team or whole engineering team depending on the scope of the tech spec
 - [x] Tech spec drafter: Facilitate that discussion.
 - [x] Other developers: Read the tech spec before arriving at the discussion.
 - [x] Other developers: Remind the tech spec drafter if a tech spec has been produced and no meeting has been scheduled.
 - [x] One or more solutions determined as viable paths forward
 - [ ] Tech spec drafter: Turn tech spec into next-step actionable tickets.
    - [ ] Write tickets as is necessary
    - [ ] Other developers provide more formal feedback as is necessary.
- [ ] Once the tech spec is finalized, open a PR to add it to the docs/tech-specs directory following [these instructions](https://github.com/department-of-veterans-affairs/caseflow/tree/master/docs/tech-specs/README.md)

### LINKED ISSUE #16490: Add EngineeringTask for use by Caseflow Engineers
From `EngineeringTask` Tech Spec #16445

## Description

Add an `EngineeringTask` class for use by Caseflow Engineers to indicate that engineers are working on the appeal.

Implement this new task for use in the Unrecognized Appellant scenario to prevent an appeal from progressing without causing a false positive by `StuckAppealsChecker`.

## Impact

There is benefit to adding a new `EngineeringTask`. It would keep Caseflow users aware of engineering work (including long-term Bat Team work) on an appeal, reduce false positives when checking for stuck appeals, and enable more accurate time-keeping when reporting on Caseflow users' time spent on an appeal.

## Out of scope

* Displaying the new task to users in Case Details, Queue page (in the on-hold tab), etc.
* Using it in an Unrecognized Appellant scenario

## Acceptance criteria
- [x] Update documentation: https://github.com/department-of-veterans-affairs/caseflow/wiki/Tasks
- [x] New task for use by Caseflow Engineers: new task can block a parent task, instructions can be added, `TimedHoldTask` can be added as a child task, task can be assigned and reassigned to engineers, etc.

## Background/context/resources
See `EngineeringTask` Tech Spec #16445

### PR DESCRIPTION
**Pull request by yoomlam**
Resolves #16490

### Description
Add a new `EngineeringTask` to keep Caseflow users aware of engineering work (including long-term Bat Team work) on an appeal, reduce false positives when checking for stuck appeals, and enable more accurate time-keeping when reporting on Caseflow users' time spent on an appeal.

Bonus: make use of `to be_valid` and `to be_invalid` more consistent for task RSpecs.

### Acceptance Criteria
- [x] Code compiles correctly
- [x] new task can block a parent task
- [x] instructions can be added
- [x] `TimedHoldTask` can be added as a child task
- [x] task can be assigned and reassigned to engineers
- [x] Add new section https://github.com/department-of-veterans-affairs/caseflow/wiki/Tasks#what-are-engineering-tasks

### Testing Plan
See RSpec.

To see changes to the UI, replicate the comment below for `judge_checkout_flow_spec.rb`.

### Reminders
* [ ] Update appeals with unrecognized appellant to use this new task.
* [ ] If desired, hide the new task to users in Case Details, Queue page (in the on-hold tab), etc.

### CODE REVIEW

**Comment by yoomlam** [comment]
For future reference, given this task tree (created by executing `EngineeringTask.create(parent: Task.find(4), appeal: case_review.appeal)` after https://github.com/department-of-veterans-affairs/caseflow/blob/20b0f667c1a1ae50f0fa05dd034a03cf49fcf42b/spec/feature/queue/judge_checkout_flow_spec.rb#L168):
```ruby
                                ┌──────────────────────────────────────────────────────────────────────┐
Appeal 1 (E 210721-1 Original)  │ ID │ STATUS    │ ASGN_BY │ ASGN_TO       │ UPDATED_AT                │
└── RootTask                    │ 1  │ on_hold   │         │ Bva           │ 2021-07-22 11:52:23 -0400 │
    ├── JudgeDecisionReviewTask │ 2  │ completed │ CSS_ID3 │ CSS_ID1       │ 2021-07-22 11:52:48 -0400 │
    │   └── AttorneyTask        │ 3  │ completed │ CSS_ID1 │ DEFAULT_USER  │ 2021-07-22 11:52:23 -0400 │
    └── QualityReviewTask       │ 4  │ on_hold   │         │ QualityReview │ 2021-07-22 11:55:41 -0400 │
        └── EngineeringTask     │ 5  │ assigned  │         │ CASEFLOW1     │ 2021-07-22 11:55:41 -0400 │
                                └──────────────────────────────────────────────────────────────────────┘
```
the Case Details page shows: 
image

Executing `Task.find(5).completed!`, then shows:
image
image

File: app/models/concerns/task_tree_render_module.rb [hunk 1]
```diff
@@ -31,7 +31,7 @@ def self.new_renderer
       ttr.config.default_atts = [:id, :status, :ASGN_BY, :ASGN_TO, :updated_at]
       ttr.config.value_funcs_hash.merge!(PRESET_VALUE_FUNCS)
       ttr.config.heading_label_template = lambda { |appeal|
-        docket = appeal.docket_name.first.titleize
+        docket = appeal.docket_name&.first&.titleize || "?"
```
Comment (yoomlam) [inline app/models/concerns/task_tree_render_module.rb:34]: Unrelated small fix: Fixes `treee` for unestablished appeals.

File: app/models/task.rb [hunk 2]
```diff
@@ -464,6 +464,10 @@ def flattened_instructions(params)
     [instructions, params.dig(:instructions).presence].flatten.compact
   end
 
+  def append_instruction(instruction)
+    update!(instructions: flattened_instructions(instructions: instruction))
+  end
+
```
Comment (yoomlam) [inline app/models/task.rb:470]: Allows engineers to more easily add to `instructions` in the Rails console.

File: spec/models/tasks/engineering_task_spec.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    context "when appeal is ready for dispatch but engineering work is needed first" do
+      let(:appeal) { create(:appeal, :at_bva_dispatch) }
+      let(:parent_task) { appeal.tasks.find_by(type: :BvaDispatchTask, assigned_to_type: "User") }
+
+      it "blocks parent BvaDispatchTask" do
+        new_task = subject
+        expect(parent_task.status).to eq "on_hold"
+        expect(new_task.status).to eq "assigned"
+      end
+    end
```
Comment (yoomlam) [inline spec/models/tasks/engineering_task_spec.rb:43]: May want to do something like this for Unrecognized Appellant work.

File: spec/models/tasks/task_shared_examples.rb [hunk 4]
```diff
@@ -16,7 +16,7 @@
     let(:parent_task) { create(:root_task) }
     it "fails to create task" do
       new_task = subject
-      expect(new_task.invalid?)
+      expect(new_task.invalid?).to eq true
```
Comment (yoomlam) [inline spec/models/tasks/task_shared_examples.rb:19]: `expect(false)` doesn't cause a failure

File: spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb [hunk 5]
```diff
@@ -26,7 +26,7 @@
 
       it "creates task with child TimedHoldTask" do
         new_task = subject
-        expect(new_task.valid?)
+        expect(new_task.valid?).to eq true
```
Comment (yoomlam) [inline spec/models/tasks/cavc_remand_processed_letter_response_window_task_spec.rb:29]: `expect(false)` doesn't cause a failure

File: spec/models/tasks/task_shared_examples.rb (hunk 4 above)
Comment (nanotone) [reply spec/models/tasks/task_shared_examples.rb:19]: Tiny nitpick: a more informative phrasing (for rspec to display failures) could be

```expect(new_task).to be_invalid```

or, with less magic,

```expect(new_task).to have_attributes(invalid?: true)```

File: app/models/tasks/engineering_task.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  def self.label
+    "Caseflow Engineering Task"
+  end
+
+  TASK_ACTIONS = [
+    Constants.TASK_ACTIONS.TOGGLE_TIMED_HOLD.to_h
+  ].freeze
+
+  def available_actions(user)
+    return [] unless user.admin?
```
Comment (lowellrex) [inline app/models/tasks/engineering_task.rb:33]: This suggestion may be introducing _yet another_ permissioning system, but would it make sense to add a `CaseflowEngineeringTeam` organization and assign this task to that organization instead of the system user. We could add everybody who is [currently listed as global or system admin](https://github.com/department-of-veterans-affairs/appeals-deployment/blob/master/ansible/vars/functions-config.yml) (the lists of global and system users are nearly identical to one another at this point) to that organization and then this task would behave incredibly similarly to existing organization tasks (and we could use the Caseflow UI to give developers admin rights instead of having to make a PR against the appeals-deployment repo and kick off an update to those functions).

Maybe a lot to consider just to get this change merged, but I figure it might be an opportunity to eat our own dog food here.

Comment (yoomlam) [reply app/models/tasks/engineering_task.rb:33]: I started to think down that path but stopped due to the coordination of action items needed, but I like the idea of dogfooding. I propose merging this PR as is and creating a plan to 
* put all the engineers into this new team
* update `functions-config.yml`
* update on-ramping instructions
* answer questions like "will this new team be shown on the Team Management page?"
* training Team Management users to prevent unintentional admin access
* etc.

Comment (lowellrex) [reply app/models/tasks/engineering_task.rb:33]: Makes sense to me!

**Review by nanotone — APPROVED** [review]
I also like the idea of punting on figuring out how to assign these tasks. So with that said, this LGTM.
